staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1201,21 +1274,32 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
 
 // Implements watch.Interface.
 func (c *cacheWatcher) Stop() {
-	c.forget()
+	c.forget(false)
 }
 
 // we rely on the fact that stopLocked is actually protected by Cacher.Lock()
 func (c *cacheWatcher) stopLocked() {
 	if !c.stopped {
 		c.stopped = true
-		close(c.done)
 		close(c.input)
 	}
+
+	// stop without draining the input channel was requested
+	// make sure the done channel is not already closed otherwise it will panic
+	if !c.drainInputBuffer && !c.isDoneChannelClosed() {
+		close(c.done)
+	}
 }
 
 func (c *cacheWatcher) nonblockingAdd(event *watchCacheEvent) bool {
+	// if the bookmarkAfterResourceVersion hasn't been seen we will try to deliver a bookmark event every second.
+	// the following check will discard a bookmark event if it is < than the bookmarkAfterResourceVersion so that we don't pollute the input channel
+	if event.Type == watch.Bookmark && !c.wasBookmarkAfterResourceVersionReceived() && event.ResourceVersion < c.bookmarkAfterResourceVersion {
+		return false // skip this bookmark event
+	}
 	select {
 	case c.input <- event:
+		c.markBookmarkAfterResourceVersionAsReceived(event)
```

Comment:
The same thing is blocking in the c.input <- call in the `add` method.

Reply:
no, bookmarks events are only added via `nonblockingAdd` method

Reply:
OK - good point. Can you please add a comment for add() method that Bookmark events are never using add?